File: bench/sort/sort.js
import { medianUs, checksumF64, printResult } from '../_lib/benchlib.js'

const N = 8192
const N_ITERS = 24
const N_RUNS = 21
const N_WARMUP = 5

const fill = (a) => {
  let s = 0x9e3779b9 | 0
  for (let i = 0; i < a.length; i++) {
    s ^= s << 13
    s ^= s >>> 17
    s ^= s << 5
    a[i] = (s >>> 0) / 4294967296
  }
}

const heapsort = (a) => {
  const n = a.length
  for (let root = (n >> 1) - 1; root >= 0; root--) {
    let i = root
    let child = 2 * i + 1
    while (child < n) {
      if (child + 1 < n && a[child] < a[child + 1]) child++
      if (a[i] >= a[child]) break
      const t = a[i]; a[i] = a[child]; a[child] = t
      i = child
      child = 2 * i + 1
    }
  }
  for (let end = n - 1; end > 0; end--) {
    const t = a[0]; a[0] = a[end]; a[end] = t
    let i = 0
    let child = 1
    while (child < end) {
      if (child + 1 < end && a[child] < a[child + 1]) child++
      if (a[i] >= a[child]) break
      const u = a[i]; a[i] = a[child]; a[child] = u
      i = child
      child = 2 * i + 1
    }
  }
}

const runKernel = (a, src) => {
  for (let it = 0; it < N_ITERS; it++) {
    for (let i = 0; i < a.length; i++) a[i] = src[i] + it
    heapsort(a)
  }
}

export let main = () => {
  const src = new Float64Array(N)
  const a = new Float64Array(N)
  fill(src)
  for (let i = 0; i < N_WARMUP; i++) runKernel(a, src)

  const samples = new Float64Array(N_RUNS)
  for (let i = 0; i < N_RUNS; i++) {
    const t0 = performance.now()
    runKernel(a, src)
    samples[i] = performance.now() - t0
  }
  printResult(medianUs(samples), checksumF64(a), N * N_ITERS, 2, N_RUNS)
}

File: bench/sort/sort.go
package main

import (
	"fmt"
	"math"
	"time"
)

const (
	n       = 8192
	nIters  = 24
	nRuns   = 21
	nWarmup = 5
)

func mix(h, x uint32) uint32 { return (h ^ x) * 0x01000193 }

func medianUs(samples []float64) int {
	for i := 1; i < len(samples); i++ {
		v := samples[i]
		j := i - 1
		for j >= 0 && samples[j] > v {
			samples[j+1] = samples[j]
			j--
		}
		samples[j+1] = v
	}
	return int(samples[(len(samples)-1)>>1] * 1000)
}

func checksumF64(xs []float64) uint32 {
	h := uint32(0x811c9dc5)
	for i := 0; i < len(xs)*2; i += 256 {
		b := math.Float64bits(xs[i>>1])
		var x uint32
		if i&1 == 0 {
			x = uint32(b)
		} else {
			x = uint32(b >> 32)
		}
		h = mix(h, x)
	}
	return h
}

func fill(xs []float64) {
	s := uint32(0x9e3779b9)
	for i := range xs {
		s ^= s << 13
		s ^= s >> 17
		s ^= s << 5
		xs[i] = float64(s) / 4294967296.0
	}
}

func heapsort(a []float64) {
	for root := (len(a) >> 1) - 1; root >= 0; root-- {
		i := root
		child := 2*i + 1
		for child < len(a) {
			if child+1 < len(a) && a[child] < a[child+1] {
				child++
			}
			if a[i] >= a[child] {
				break
			}
			a[i], a[child] = a[child], a[i]
			i = child
			child = 2*i + 1
		}
	}
	for end := len(a) - 1; end > 0; end-- {
		a[0], a[end] = a[end], a[0]
		i := 0
		child := 1
		for child < end {
			if child+1 < end && a[child] < a[child+1] {
				child++
			}
			if a[i] >= a[child] {
				break
			}
			a[i], a[child] = a[child], a[i]
			i = child
			child = 2*i + 1
		}
	}
}

func runKernel(a, src []float64) {
	for it := 0; it < nIters; it++ {
		f := float64(it)
		for i := range a {
			a[i] = src[i] + f
		}
		heapsort(a)
	}
}

func main() {
	src := make([]float64, n)
	a := make([]float64, n)
	fill(src)
	for i := 0; i < nWarmup; i++ {
		runKernel(a, src)
	}
	samples := make([]float64, nRuns)
	for i := 0; i < nRuns; i++ {
		t0 := time.Now()
		runKernel(a, src)
		samples[i] = float64(time.Since(t0).Nanoseconds()) / 1e6
	}
	fmt.Printf("median_us=%d checksum=%d samples=%d stages=%d runs=%d\n",
		medianUs(samples), checksumF64(a), n*nIters, 2, nRuns)
}
